### DISCUSSION TITLE
[Feature request] - New column and field types for ```spatie/laravel-data```

### DISCUSSION THREAD

**Discussion by susanu** [Laravel-Backpack/community-forum#1116] [Ideas]
Hi guys,

I need to edit and display data attributes, but there is no out of the box solution for this.
The closest to achieve this is with ```repeatable``` with ```'min_rows' => 1```, ```'max_rows' => 1``` and ```'value' => fn($entry) => [$entry->data->toArray()]```.
This solution is very hacky and is not sustainable.
Is there any chance we would get support for data attributes or key => value arrays?
We can display this in a vertical table, example: https://www.w3schools.com/hTml/tryit.asp?filename=tryhtml_table_headers

Thank you!

**Comment by susanu** [comment]
I made a quick concept, showing only for column because for field is just a repeatable with 1 row.

Setup, ```shipping_address``` is casted to ```Address```:
```
<?php

use Spatie\LaravelData\Data;

class Address extends Data
{
    public function __construct(
        public string $first_name,
        public string $last_name,
    ) {

    }
}
```

Column setup:
```
CRUD::column('shipping_address')
    ->type('vertical_table')
    ->columns([
        'first_name' => 'First name',
        'last_name' => 'Last name'
    ]);
```

Column code: ```vertical_table.blade.php```
```
@php
	$column['value'] = $column['value'] ?? data_get($entry, $column['name']);
    $column['columns'] = $column['columns'] ?? ['value' => 'Value'];
	$column['escaped'] = $column['escaped'] ?? true;

    if($column['value'] instanceof \Closure) {
        $column['value'] = $column['value']($entry);
    }

	// if this attribute isn't using attribute casting, decode it
	if (is_string($column['value'])) {
	    $column['value'] = json_decode($column['value'], true);
    }

    if(is_a($column['value'], \Illuminate\Contracts\Support\Arrayable::class)) {
        $column['value'] = $column['value']->toArray();
    }

	// always work with arrays in the html, so if it is an object, get an array back from it.
	if(is_object($column['value'])) {
		$column['value'] = (array)$column['value'];
	}
@endphp

<span>
    @if (!empty($column['value']) && count($column['columns']))

    @includeWhen(!empty($column['wrapper']), 'crud::columns.inc.wrapper_start')

    <table class="table table-bordered table-condensed table-striped m-b-0">
		<tbody>
			@foreach ($column['columns'] as $tableColumnKey => $tableColumnLabel)
			<tr>
                <th>
                    {{ $tableColumnLabel }}
                </th>
				<td>
                    @if($column['escaped'])
                        {{ $column['value'][$tableColumnKey] ?? '' }}
                    @else
                        {!! $column['value'][$tableColumnKey] ?? '' !!}
                    @endif
                </td>
			</tr>
			@endforeach
		</tbody>
    </table>

    @includeWhen(!empty($column['wrapper']), 'crud::columns.inc.wrapper_end')

    @else

    {{ $column['default'] ?? '-' }}

	@endif
</span>
```

Result
![image](https://github.com/user-attachments/assets/e93c4f00-1f87-42bd-b121-e1de0cc4b0d9)

**Comment by pxpm** [comment]
Hey @susanu thanks a lot for the suggestion and for providing the example. 

If it works with repeatable, wouldn't it be easier to just check in repeatable if the value is a `Spatie\LaravelData\Data`, in case it is, we do the modifications you said? `min/max rows` and transforming the value into an array ? 

So you would do:
```php
CRUD::column('my_spatie_data_source')->type('repeatable');
```
Let me know what you think. 👍

**Comment by susanu** [comment]
To be fair, i don't like this approach with repeatable going forward.
When i think of repeatable, first thing that crosses my mind is multiple items, not a single array/object.

In my example, yes, it works for columns, but for fields i need to hook ```retrieved``` and ```saving``` events for the model to adjust the value of the repeatable, which is not something i want to do long term.

As of now, there is a ```json``` column type but there is no field type.
What do you think of creating a new ```json``` field and will support json objects and data objects.
If the developer sets ```subfields```, show a ```repeatable``` type field with only 1 row.
Otherwise, show a textarea with the raw json.

Also changing the current ```json``` column to display a vertical table or something similar if the developer defines the ```columns``` attribute.
Otherwise, leave the current behavior.

What do you think?

**Comment by susanu** [comment]
Hi @pxpm,

What do you think?
I can help you with this if you decide to make it a thing.

Thanks!

**Reply by pxpm** [reply]
Hey @susanu sorry I didn't had time to look at this again. 

This needs some thinking before we decide to add/change columns/fields. 
Regarding the `retrieved/saving` we can do that part in repeatable, like I told you I don't mind having a condition in repeatable that handle special data, we already have some fields like that, we call then "switchboard fields", basically are fields that "prepare and include" other fields, but they don't have "logic" on them. 

In this case, we could have a `json` field, that is just a switchboard to either `repeatable` or `textarea`, without any logic in it, so in your code it would be `CRUD::field('smt')->type('json')`, but under the hood we would "prepare and handle" the repeatable stuff. 

Not 100% sure of how it could work, as I told you, I didn't spent much time thinking about this yet, nor trying to come up with a POC. 

I know spatie/laravel-data is a widely used package and that's why I am willing to give it a shot, but this is not on our priority list. 

To avoid messing with your model, you can define the `retrieved` and `saving` events directly on the field: https://backpackforlaravel.com/docs/6.x/crud-operation-update#use-events-in-your-field-definition

Cheers 🙏
